pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,327 @@
+	igmpSnooper  *IGMPSnooper
+	ofClient     openflow.Client
+	groupEventCh chan *mcastGroupEvent
+	staleGroupCh chan net.IP
+	groupCache   cache.Indexer
+	ctx          context.Context
+}
+
+func (w *worker) processGroupEvent(stopCh <-chan struct{}) {
+	for {
+		select {
+		case e := <-w.groupEventCh:
+			obj, ok, _ := w.groupCache.GetByKey(e.group.String())
+			switch e.eType {
+			case groupJoin:
+				if !ok {
+					w.addGroupMemberStatus(e)
+				} else {
+					w.updateGroupMemberStatus(obj, e)
+				}
+			case groupLeave:
+				if ok {
+					w.updateGroupMemberStatus(obj, e)
+				}
+			}
+		case group := <-w.staleGroupCh:
+			obj, exists, _ := w.groupCache.GetByKey(group.String())
+			if !exists {
+				continue
+			}
+			status := obj.(*GroupMemberStatus)
+			membersCount := len(status.localMembers)
+			diff := time.Now().Sub(status.lastProbe).Seconds()
+			if membersCount == 0 || diff > mcastGroupTimeout.Seconds() {
+				// Remove the multicast flow entry if no local Pod has joined the group.
+				if err := w.ofClient.UninstallMulticastFlow(group); err != nil {
+					klog.ErrorS(err, "Failed to uninstall multicast flows", "group", group.String())
+				}
+				// TODO: Remove group from multicast routing entries.
+				w.groupCache.Delete(status)
+				klog.InfoS("Removed multicast group from cache after all members left", "group", group.String())
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+// addGroupMemberStatus adds the new group into groupCache.
+func (w *worker) addGroupMemberStatus(e *mcastGroupEvent) error {
+	if err := w.ofClient.InstallMulticastFlow(e.group); err != nil {
+		return err
+	}
+	status := &GroupMemberStatus{
+		group:        e.group,
+		lastProbe:    e.time,
+		localMembers: sets.NewString(e.iface.InterfaceName),
+	}
+	// TODO: add multicast routing entries for the new added multicast group before add status into groupCache.
+	w.groupCache.Add(status)
+	klog.InfoS("Added new multicast group to cache", "group", e.group, "interface", e.iface.InterfaceName)
+	return nil
+}
+
+// updateGroupMemberStatus updates the group status in groupCache. If a "join" message is sent from an existing member,
+// only updates the lastProbe time. If a "join" message is sent from an "unknown" member, updates the lastProbe time and
+// adds the new member into the group's local member set. If a "leave" message is sent from an existing member, remove it from the
+// group's local member set, and if the member is the last one in local cache, a query message on the group is sent out
+// to check if there are still local members in the group.
+func (w *worker) updateGroupMemberStatus(obj interface{}, e *mcastGroupEvent) error {
+	status := obj.(*GroupMemberStatus)
+	status.mutex.Lock()
+	defer status.mutex.Unlock()
+	exist := status.localMembers.Has(e.iface.InterfaceName)
+	switch e.eType {
+	case groupJoin:
+		status.lastProbe = e.time
+		if !exist {
+			// TODO: process the new member in the multicast routing entries if necessary.
```

Comment:
It is a TODO leaving for Multicast routing entry handler which is in another PR.

Reply:
Maybe I should remove such TODOs in my PR...